# Curso de Introducción a la Ciencia de Datos

**Ejercicio SF Salaries**
Utilizaremos el `SF Salaries Dataset` Kaggle. Simplemente siga y complete las tareas que se detallan a continuación en negrita. Las tareas serán cada vez más difíciles a medida que avance.

La idea principal es aplicar pandas para organizar informacion, manipularla, buscar información en el dataset


**Importe pandas como pd. e imprima la version que esta usando**

In [230]:
# Copie el código aca

import pandas as pd

print(pd.__version__)

3.0.2


**leer el archivo Salaries.csv en un dataframe llamado salarios**

se encuentra en esta url y se puede leer directamente con pandas

https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv


In [231]:
# Copie el código aca
#Leemos el archivo csv con pandas

salarios = pd.read_csv("https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv")
salarios.head()


,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Agency,Status
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,NaN,San Francisco,NaN
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,NaN,San Francisco,NaN
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,NaN,San Francisco,NaN
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,NaN,San Francisco,NaN
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,NaN,San Francisco,NaN


**Use el método `.info()` para encontrar cuantas Columnas de atributos hay y que tipo de datos tiene cada una.**

In [232]:
# Copie el código aca
salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  str    
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), str(3)
memory usage: 14.7 MB


- Segun la información del dataframe su tamaño es de *memory usage: 14.7+ MB*

- las columnas **Notes** y **Status** solamente tienen valores *NaN*, verifique que dice **0 non-null** osea 0 valores son diferentes a *NaN*
entonces todos los valores son *NaN* en esas columnas

**Elimine las columnas *Notes* y *Status* Recuerde usar el método de forma permanente (inplace), y luego visualice cuantas filas y columnas tiene luego de eliminar las columnas**

In [233]:
# Copie el código aca
#Eliminamos la columna notes
salarios.drop(columns=["Notes"], inplace=True)

#Eliminamos la columna Status
salarios.drop(columns=["Status"], inplace=True)



In [234]:
salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


**¿Cual es la nueva información del dataframe?**

In [235]:
# Copie el código aca
salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


- Se puede verificar que al borrar las columnas que no tenian solo *NaN* se disminuye el tamaño del dataframe a: *memory usage: 12.5+ MB*

**Observe los primeros 5 registros del DataFrame.**

In [236]:
# Copie el código aca

salarios.head(5)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,San Francisco
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,San Francisco
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,San Francisco
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,San Francisco
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,San Francisco


**Observe los ultimos 5 registros del DataFrame.**

In [237]:
# Copie el código aca

salarios.tail(5)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
148649,148650,Roy I Tillery,Custodian,0.0,0.0,0.00,0.0,0.00,0.00,2014,San Francisco
148650,148651,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148651,148652,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148652,148653,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148653,148654,Joe Lopez,"Counselor, Log Cabin Ranch",0.0,0.0,-618.13,0.0,-618.13,-618.13,2014,San Francisco


**¿Cuantas filas y columnas tiene el dataframe?**

In [238]:
# Copie el código aca

salarios.shape

(148654, 11)

**Observe los 10 registros que estan desde el indice 3000**

In [239]:
# Copie el código aca

#Observamos los 10 registros que estan desde el indice 3000
salarios[3000:3010]

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
3000,3001,WILLIAM BOYLE,FIREFIGHTER,105934.65,18530.98,16389.60,NaN,140855.23,140855.23,2011,San Francisco
3001,3002,ALAN LIN,ELECTRONIC MAINTENANCE TECHNICIAN,98965.80,41504.40,380.00,NaN,140850.20,140850.20,2011,San Francisco
3002,3003,WILLIAM CONLEY,SERGEANT III (POLICE DEPARTMENT),130457.78,3131.42,7256.39,NaN,140845.59,140845.59,2011,San Francisco
3003,3004,ROBERT MCDONALD,"LIEUTENANT, FIRE DEPARTMENT",123104.99,1833.56,15897.68,NaN,140836.23,140836.23,2011,San Francisco
3004,3005,MICHAEL COSTELLO,POLICE OFFICER III,112401.08,7957.74,20466.93,NaN,140825.75,140825.75,2011,San Francisco
3005,3006,EDNA HO,MANAGER V,140821.37,0.00,0.00,NaN,140821.37,140821.37,2011,San Francisco
3006,3007,RODRIGO CONEL,REGISTERED NURSE,126120.00,3981.78,10714.21,NaN,140815.99,140815.99,2011,San Francisco
3007,3008,VICTOR HURTADO,FIREFIGHTER,105934.67,13112.76,21763.50,NaN,140810.93,140810.93,2011,San Francisco
3008,3009,DONALD LANDRY,TRANSIT POWER LINE WORKER,92861.50,39789.32,8135.86,NaN,140786.68,140786.68,2011,San Francisco
3009,3010,FRANK LEE,"INSPECTOR III, (POLICE DEPARTMENT)",130481.56,725.40,9579.58,NaN,140786.54,140786.54,2011,San Francisco



La columna *Benefits* tiene valores `NaN`, ¿Cuantos NaN tiene la columna *Benefits*? (TIP: Seleccione la Columna y vefique cuantos Na tiene `.isna()` luego sumelos `.sum()` )

In [240]:
# Copie el código aca
print(salarios["Benefits"].isna().sum())

36163


Al parecer los valores faltantes en la columna Benefits es por que no se le pagan beneficios entonces:

**Reemplace los valores *NaN* de la Columna *Benefits* por el valor de cero (Recuerde usar el método de forma inplace) y mostrar los primeros 10 elementos de la lista**

In [241]:
# Copie el código aca

#Reemplazamos los valores nulos de la columna Benefits por "0"

salarios["Benefits"] = salarios["Benefits"].fillna(0)

salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          148654 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


In [242]:
salarios.head(10)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,0.0,567595.43,567595.43,2011,San Francisco
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,0.0,538909.28,538909.28,2011,San Francisco
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,0.0,335279.91,335279.91,2011,San Francisco
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,0.0,332343.61,332343.61,2011,San Francisco
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,0.0,326373.19,326373.19,2011,San Francisco
5,6,DAVID SULLIVAN,ASSISTANT DEPUTY CHIEF II,118602.00,8601.00,189082.74,0.0,316285.74,316285.74,2011,San Francisco
6,7,ALSON LEE,"BATTALION CHIEF, (FIRE DEPARTMENT)",92492.01,89062.90,134426.14,0.0,315981.05,315981.05,2011,San Francisco
7,8,DAVID KUSHNER,DEPUTY DIRECTOR OF INVESTMENTS,256576.96,0.00,51322.50,0.0,307899.46,307899.46,2011,San Francisco
8,9,MICHAEL MORRIS,"BATTALION CHIEF, (FIRE DEPARTMENT)",176932.64,86362.68,40132.23,0.0,303427.55,303427.55,2011,San Francisco
9,10,JOANNE HAYES-WHITE,"CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)",285262.00,0.00,17115.73,0.0,302377.73,302377.73,2011,San Francisco


**¿Cual es la nueva información del dataframe?**

In [243]:
# Copie el código aca

salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          148654 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


**Cuantos valores unicos hay en las columnas categoricas?**

In [244]:
# Copie el código aca
#Variables categoricas
print(salarios["EmployeeName"].nunique())
print(salarios["JobTitle"].nunique())
print(salarios["Agency"].nunique())

110811
2159
1


- la columna **Agency** solamente tiene un valor unico, entonces no aporta informacion, por lo tanto se puede eliminar

**Elimine la columna *Agency* del dataframe**

In [245]:
# Copie el código aca
salarios.drop(columns=["Agency"], inplace = True)

In [246]:
salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          148654 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
dtypes: float64(6), int64(2), str(2)
memory usage: 11.3 MB


**Cual es el promedio de BasePay ?**

In [247]:
# Copie el código aqui
print(f"El promedio de BasePay es: {salarios["BasePay"].mean():.4f}")


El promedio de BasePay es: 66325.4488


**Cual es la cantidad mas alta de OvertimePay en el dataset ?**

In [248]:
# Copie el código aca
print(f"El maximo OvertimePay es: {salarios["OvertimePay"].max():.2f}")

El maximo OvertimePay es: 245131.88


**Realice una descripcion estadistica de las columnas numericas?**

In [249]:
# Copie el código aca
#Se realiza una descripcion estadistica para BasePay, overtimePay, OtherPay, Benefits, TotalPay, TotalPayBenefits, y Year
for column in ["BasePay", "OvertimePay", "OtherPay", "Benefits", "TotalPay", "TotalPayBenefits", "Year"]:
    print(f"Descrpcion estadistica de {column}")
    print(salarios[column].describe())
    print("\n")


Descrpcion estadistica de BasePay
count    148045.000000
mean      66325.448840
std       42764.635495
min        -166.010000
25%       33588.200000
50%       65007.450000
75%       94691.050000
max      319275.010000
Name: BasePay, dtype: float64


Descrpcion estadistica de OvertimePay
count    148650.000000
mean       5066.059886
std       11454.380559
min          -0.010000
25%           0.000000
50%           0.000000
75%        4658.175000
max      245131.880000
Name: OvertimePay, dtype: float64


Descrpcion estadistica de OtherPay
count    148650.000000
mean       3648.767297
std        8056.601866
min       -7058.590000
25%           0.000000
50%         811.270000
75%        4236.065000
max      400184.250000
Name: OtherPay, dtype: float64


Descrpcion estadistica de Benefits
count    148654.000000
mean      18924.232839
std       17165.279495
min         -33.890000
25%           0.000000
50%       23214.020000
75%       33468.980000
max       96570.660000
Name: Benefits, dtype

**En este dataset las variables `Id` y `Year` representan variables categorica, convertir estas columnas a su formato correcto**

In [250]:
# Copie el código aca
#Para Year
salarios["Year"] = salarios["Year"].astype(str)

In [251]:
#Para Id
salarios["Id"] = salarios["Id"].astype(str)

**la variable `Year` es categorica ordinal, primero detecte cuales son los valores unicos que tiene la columna**

In [252]:
# Copie el código aca
salarios["Year"].unique()

<StringArray>
['2011', '2012', '2013', '2014']
Length: 4, dtype: str

**Apartir de la información anterior convierta la columna `Year` en variable categorica ordinal y visualice la columna**

In [253]:
salarios["Year"].describe()

count     148654
unique         4
top         2014
freq       38123
Name: Year, dtype: object

In [270]:
salarios["Year"] = pd.Categorical(
    salarios["Year"],
    categories=["2011", "2012", "2013", "2014"],
    ordered=True
)

salarios["Year"].describe()
salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype   
---  ------            --------------   -----   
 0   Id                148654 non-null  str     
 1   EmployeeName      148654 non-null  str     
 2   JobTitle          148654 non-null  str     
 3   BasePay           148045 non-null  float64 
 4   OvertimePay       148650 non-null  float64 
 5   OtherPay          148650 non-null  float64 
 6   Benefits          148654 non-null  float64 
 7   TotalPay          148654 non-null  float64 
 8   TotalPayBenefits  148654 non-null  float64 
 9   Year              148654 non-null  category
dtypes: category(1), float64(6), str(3)
memory usage: 10.3 MB


**Realice nuevamente una descripcion estadistica de las columnas numericas?**

In [275]:
# Copie el código aca
for column in ["BasePay", "OvertimePay", "OtherPay", "Benefits", "TotalPay", "TotalPayBenefits"]:
    print(f"Columna: {column}")
    print(f"Descripcion: \n{salarios[column].describe()}")
    print("\n")
    

Columna: BasePay
Descripcion: 
count    148045.000000
mean      66325.448840
std       42764.635495
min        -166.010000
25%       33588.200000
50%       65007.450000
75%       94691.050000
max      319275.010000
Name: BasePay, dtype: float64


Columna: OvertimePay
Descripcion: 
count    148650.000000
mean       5066.059886
std       11454.380559
min          -0.010000
25%           0.000000
50%           0.000000
75%        4658.175000
max      245131.880000
Name: OvertimePay, dtype: float64


Columna: OtherPay
Descripcion: 
count    148650.000000
mean       3648.767297
std        8056.601866
min       -7058.590000
25%           0.000000
50%         811.270000
75%        4236.065000
max      400184.250000
Name: OtherPay, dtype: float64


Columna: Benefits
Descripcion: 
count    148654.000000
mean      18924.232839
std       17165.279495
min         -33.890000
25%           0.000000
50%       23214.020000
75%       33468.980000
max       96570.660000
Name: Benefits, dtype: float64




**¿Cuál es el título de trabajo de JOSEPH DRISCOLL? Nota: Utilice mayúsculas, de lo contrario, puede obtener una respuesta que no coincida.**

In [281]:
# Copie el código aca

salarios.head(5)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,0.0,567595.43,567595.43,2011
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,0.0,538909.28,538909.28,2011
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,0.0,335279.91,335279.91,2011
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,0.0,332343.61,332343.61,2011
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,0.0,326373.19,326373.19,2011


In [291]:
#Consultamos cuantos registros hay para el empleado JOSEPH DRISCOLL
(salarios["EmployeeName"] == "JOSEPH DRISCOLL").sum()

np.int64(1)

In [287]:
#buscamos el trabajo de JOSEPH DRISCOLL
print(f"El trabajo de JOSEPH DRISCOLL es: {salarios.loc[salarios["EmployeeName"] == "JOSEPH DRISCOLL", "JobTitle"].values[0]}")

El trabajo de JOSEPH DRISCOLL es: CAPTAIN, FIRE SUPPRESSION


**¿Cuanto gana JOSEPH DRISCOLL (incluyendo beneficios)?**

In [289]:
# Copie el código aca
print(f"El salarios de JOSEPH DRISCOLL es: {salarios.loc[salarios["EmployeeName"] == "JOSEPH DRISCOLL", "TotalPayBenefits"].values[0]}")

El salarios de JOSEPH DRISCOLL es: 270324.91


**¿Cual es el nombre de la persona con mejor pago (incluyendo beneficios)?**

In [296]:
# Copie el código aca
#Primero debemos encontarar el valor max de totalpaybenefits
max_totalpaybenefits = salarios["TotalPayBenefits"].max()
print(f"El valor maximo pagado es: {max_totalpaybenefits:.2f}")

#Averiguamos cuantas veces se repite ese valor maximo
cantidad_max_totalpaybenefits = (salarios["TotalPayBenefits"] == max_totalpaybenefits).sum()
print(f"El valor maximo de TotalPayBenefits se repite {cantidad_max_totalpaybenefits} veces")

#Miramos quien es la persona que se gana ese dinero
print(f"El empleado que mas gana es: {salarios.loc[salarios["TotalPayBenefits"] == max_totalpaybenefits, "EmployeeName"].values[0]}")

El valor maximo pagado es: 567595.43
El valor maximo de TotalPayBenefits se repite 1 veces
El empleado que mas gana es: NATHANIEL FORD


**¿Cual es el nombre de la persona peor paga (incluyendo beneficios)? ¿Notan algo extraño sobre cuánto le pagan?**

In [297]:
# Copie el código aca
#Primero debemos encontarar el valor min de totalpaybenefits
min_totalpaybenefits = salarios["TotalPayBenefits"].min()
print(f"El valor minimo pagado es: {min_totalpaybenefits:.2f}")

#Averiguamos cuantas veces se repite ese valor minimo
cantidad_min_totalpaybenefits = (salarios["TotalPayBenefits"] == min_totalpaybenefits).sum()
print(f"El valor minimo de TotalPayBenefits se repite {cantidad_min_totalpaybenefits} veces")

#Miramos quien es la persona que se gana ese dinero
print(f"El empleado que menos gana es: {salarios.loc[salarios["TotalPayBenefits"] == min_totalpaybenefits, "EmployeeName"].values[0]}")

El valor minimo pagado es: -618.13
El valor minimo de TotalPayBenefits se repite 1 veces
El empleado que menos gana es: Joe Lopez


**Cuantos registros tienen un valor de TotalPayBeneficts menor que cero**

In [299]:
# Copie el código aca
#Cantidad de totalpaybenefits menores a 0
print(f"La cantidad de registros menores que cero son: {(salarios["TotalPayBenefits"] < 0).sum()}")

La cantidad de registros menores que cero son: 4


**¿Cual es el promedio (mean) BasePay de todos los empleados por año? los años van entre (2011-2014)?** (Tip: Usar groupby)

In [303]:
# Copie el código aca
#Vamos a averiguar cual es el promedio de BasePay de todos los empleados por año
promedio_basepay_por_año = salarios.groupby("Year")["BasePay"].mean()
print("El promedio de BasePay por años es:")
print(promedio_basepay_por_año)

El promedio de BasePay por años es:
Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64


**¿Cuantos titulos de trabajos (JobTitle) unicos existen?**

In [306]:
# Copie el código aca
print(f"Los trabajos unicos son: {salarios["JobTitle"].nunique()}")

Los trabajos unicos son: 2159


**¿Cuales son los 5 trabajos mas comunes?**

In [309]:
#Los 5 trabajos mas comunes son
print(f"Los 5 trabajos mas comunes son: \n{salarios["JobTitle"].value_counts().head(5)}")

Los 5 trabajos mas comunes son: 
JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64


**¿Cuántos puestos de trabajo Unicos existieron en 2013? (Ejemplo, los títulos de trabajo con solo una ocurrencia en 2013?)**

In [310]:
#Trabajos unicos solo en el 2013
trabajos_2013 = salarios[salarios["Year"] == "2013"]["JobTitle"].nunique()
print(f"Los trabajos unicos en el 2013 son: {trabajos_2013}")

Los trabajos unicos en el 2013 son: 1051


**¿Cuántas personas tienen la palabra Chief en el título de su trabajo?** Apoyese en los métodos para manipular strings y en el método `.apply()`

In [314]:
# Copie el código aca
#Las personas que tienen en su titulo de trabajo la palabra "chief" son:
print(f"Las personas que tienen en su titulo de trabajo la palabra Chief son: \n{(salarios["JobTitle"].apply(lambda x: "CHIEF" in str(x).lower())).sum()}")

Las personas que tienen en su titulo de trabajo la palabra Chief son: 
0


Este código es valido, se recomienda hacer el código mas simple y agregar typing y documentación.

**Bono: ¿Existe una correlación entre la longitud del nombre del Título del trabajo y el Salario?**

crea una nueva columna donde se guarde la longitud del titulo del trabajo. tip: usar el método `.apply()`

In [319]:
# Copie el código aca
#Creamos la nueva columns llamada longitud, que nos dice la longitud del nombre del titulo de trabajo de cada empleado
salarios["longitud"] = salarios["JobTitle"].apply(lambda x: len(str(x)))

salarios.head(5)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,longitud
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,0.0,567595.43,567595.43,2011,46
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,0.0,538909.28,538909.28,2011,31
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,0.0,335279.91,335279.91,2011,31
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,0.0,332343.61,332343.61,2011,36
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,0.0,326373.19,326373.19,2011,44


In [321]:
# Correlación entre Longitud del titulo de trabajo y el salario total
correlacion = salarios["longitud"].corr(salarios["TotalPayBenefits"])
print(f"La correlación entre la longitud del titulo de trabajo y el salario total es: {correlacion:.4f}")

La correlación entre la longitud del titulo de trabajo y el salario total es: -0.0369



*Al ser una correlación negativa (-0.0369) quiere decir que no existe una relación entre la longitud del titulo de trabajo y el salario total.*